lib/plugins.ts

```diff
@@ -490,38 +482,29 @@ export class Plugins {
     }
     return deferred.promise;
   }
-}
-
-/**
- * Generates the handler for a plugin function (e.g. the setup() function)
- *
- * @param {string} funName The name of the function to make a handler for
- * @param {PromiseType} promiseType The type of promise (WebDriver or Q) that
- *    should be used
- * @param {boolean=} failReturnVal The value that the function should return if
- *     the plugin crashes
- *
- * @return {Function} The handler
- */
-function pluginFunFactory(
-    funName: string, promiseType: PromiseType,
-    failReturnVal?: boolean): Function {
-  return function() {
-    var promises: any[] = [];
-    var args = arguments;
-    var self: Plugins = this;
-
-    self.pluginObjs.forEach((pluginObj: ProtractorPlugin) => {
-      if ((<any>pluginObj)[funName]) {
-        promises.push(self.safeCallPluginFun(
-            pluginObj, funName, args, promiseType, failReturnVal));
-      }
-    });
 
-    if (promiseType == PromiseType.Q) {
-      return Q.all(promises);
-    } else {
-      return webdriver.promise.all(promises);
-    }
-  };
+  /**
+   * Generates the handler for a plugin function (e.g. the setup() function)
+   *
+   * @param {string} funName The name of the function to make a handler for
+   * @param {PromiseType} promiseType The type of promise (WebDriver or Q) that should be used
+   * @param {boolean=} failReturnVal The value that the function should return if the plugin crashes
+   *
+   * @return The handler
+   */
+  private pluginFunFactory(funName: string, promiseType: PromiseType.Q, failReturnVal?: boolean):
```

Comment:
The most elegant solution would be not to use the `PromiseType` enum at all, and instead make the function generic and pass the promise implementation itself (`q` or `webdriver.promise`) instead of an enum value. It'd look like this:
```ts
private pluginFunFactory<T extend typeof q | typeof webdriver.promise>(
  funName: string, promiseImplementation: T, failReturnVal?: boolean) {
...
}
```
Unfortunately, this doesn't work because of this: https://github.com/Microsoft/TypeScript/issues/11940